solidity/nft_token.sol

```diff
@@ -0,0 +1,254 @@
+import "./library/spl_token.sol";
+import "./library/system_instruction.sol";
+
+struct Position {
+    address whirlpool;      // 32
+    address position_mint;  // 32
+    uint128 liquidity;      // 16
+    int32 tick_lower_index; // 4
+    int32 tick_upper_index; // 4
+}
+
+@program_id("DPvSokZKEAYqh6jjmGDUyd4L881uAWg4rFVeA6rec2fr")
+contract nft_token {
+    address public orca;
+    address public whirlpool;
+    address public pdaProgram;
+    address public mintErc20;
+    address public pdaERC20Account;
+    // TODO: Change for the latter one in production
+    uint64 public pdaHeader = 0x55bdfe33;//0xd0f7407ae48fbcaa;
+    // PDA seed
+    bytes public constant pdaProgramSeed = "pdaProgram";
+    // PDA bump
+    bytes1 public pdaBump;
+    int32 public constant min_tick_lower_index = -443632;
+    int32 public constant max_tick_lower_index = 443632;
+    //address constant tokenProgram = address"TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA";
+
+    uint64 public numTokenAccounts;
+
+    mapping(address => uint64) public mapTokenAccountLiquidity;
+    address[type(uint32).max] public tokenAccounts;
+
+    @space(10000)
+    @payer(payer)
+    @seed("pdaProgram")
+    constructor(
+        address _orca,
+        address _whirlpool,
+        address _mintErc20,
+        address _pdaERC20Account,
+        @bump bytes1 _bump
+    ) {
+        orca = _orca;
+        whirlpool = _whirlpool;
+        mintErc20 = _mintErc20;
+        pdaERC20Account = _pdaERC20Account;
+
+        // Independently derive the PDA address from the seeds, bump, and programId
+        (address pda, bytes1 bump) = try_find_program_address(["pdaProgram"], type(nft_token).program_id);
+
+        // Verify that the bump passed to the constructor matches the bump derived from the seeds and programId
+        if (bump != _bump) {
+            revert("Invalid bump");
+        }
+
+        pdaBump = bump;
+        pdaProgram = pda;
+    }
+
+    function getPositionData(address positionDataAccount, address positionMint) public view returns (Position position) {
```

Comment:
Presumably the liquidity pair needs to exist before we can deploy this?

Reply:
Yes, however we will revise this check in a next iteration.